// @netlify/build-info/lib/node/get-build-info.js
import { listFrameworks } from '@netlify/framework-info';
import { report } from '../metrics.js';
import { Project } from '../project.js';
import { NodeFS } from './file-system.js';
/** A noop logger that is used to not log anything (we use the stdout for parsing the json output) */
export class NoopLogger {
    debug() {
        /** noop */
    }
    log() {
        /** noop */
    }
    error() {
        /** noop */
    }
    info() {
        /** noop */
    }
    warn() {
        /** noop */
    }
}
/** Get the build info object that is used inside buildbot */
export async function getBuildInfo(config = { featureFlags: {} }) {
    const fs = new NodeFS();
    // prevent logging in output as we use the stdout to capture the json
    fs.logger = new NoopLogger();
    const project = new Project(fs, config.projectDir, config.rootDir)
        .setBugsnag(config.bugsnagClient)
        .setEnvironment(process.env)
        .setNodeVersion(process.version);
    const info = {};
    if (config.featureFlags?.build_info_new_framework_detection) {
        info.frameworks = (await project.detectFrameworksInPath(project.baseDirectory)) || [];
    }
    else {
        try {
            // if the framework  detection is crashing we should not crash the build info and package-manager detection
            info.frameworks = (await listFrameworks({ projectDir: project.baseDirectory }));
        }
        catch (error) {
            report(error, { client: config.bugsnagClient });
            info.frameworks = [];
        }
    }
    info.settings = await project.getBuildSettings();
    info.langRuntimes = await project.detectRuntime();
    // some framework detection like NX can update the workspace in the project so assign it later on
    info.jsWorkspaces = project.workspace;
    info.buildSystems = project.buildSystems;
    info.packageManager = project.packageManager;
    return info;
}
